src/ipc/ipc4/handler.c

```diff
@@ -999,9 +1030,10 @@ void ipc_cmd(struct ipc_cmd_hdr *_hdr)
 	if (err)
 		tr_err(&ipc_tr, "ipc4: %d failed err %d", target, err);
 
+	ipc_platform_ack_cmd();
```

Comment:
One point is that: FW will access memory window to process cmd, I am not sure ack it before memory access will cause any issue or not

Reply:
Hrm, true. I think the definition of the ACK is not that clear.
Is it means that DSP received the message and it is started to process it, new message should not be sent until the reply is sent back.
or
DSP received the message and it got the header and payload (if any), host can send a new message. A reply will be sent later with the results.

The kernel implementation is the first one, we serialize messages and never have two tx flying at the same time and the ack is kind of just ignored, we ack the ack and that's it. I think ROM messages might be using the ack as a reply (as IPC3 does), but we only send ROM message once on boot.

Reply:
@RanderWang what happens in the case of D3? Typically, we'd avoid sending the ack until after we've powered off the SRAMs but in this case you wouldnt. Could it possibly cause some issues?

Reply:
> Hrm, true. I think the definition of the ACK is not that clear. Is it means that DSP received the message and it is started to process it, new message should not be sent until the reply is sent back. or DSP received the message and it got the header and payload (if any), host can send a new message. A reply will be sent later with the results.

I think its a hybrid version of these two methods:  ACK when DSP received the message and it got the header and payload (if any),  new message should not be sent until the reply is sent back.

Reply:
> In my PR we ack first then power off SRAM

@RanderWang I can tell that from the code. But isnt that racy? What if the host powers down the DSP before the firmware can power of the SRAMs

Reply:
for IPC4,  can host powers down DSP with reply msg received ?

Reply:
Currently both Linux & windows driver wait for the reply msg of set_dx. Error without reply msg

Reply:
@ranj063, this is what happens with SOF.IPC4 firmware:
```
sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x47000000|0x0: MOD_SET_DX [data size: 8]
cnl_ipc4_irq_thread: DSP ack + msg
sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x67000000|0x0: MOD_SET_DX
sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x47000000|0x0: MOD_SET_DX [data size: 8]
```
and this is what supposed to happen:
```
sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x47000000|0x0: MOD_SET_DX [data size: 8]
cnl_ipc4_irq_thread: DSP ack + msg
sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x67000000|0x0: MOD_SET_DX
sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x47000000|0x0: MOD_SET_DX [data size: 8]
```
With SOF we do not get an ack for the MOD_SET_DX currently because it is set _after_ the reply is sent currently and this is wrong.
The firmware 1st must ack the reception of a message, process it and send a reply. All messages, including the MOD_SET_DX must have an ack followed by a reply.

Hrm: https://github.com/thesofproject/sof/pull/5853

Reply:
Even with #5853 we are missing the ack since it is done in `ipc_platform_complete_cmd()` which is not called due to power off, so the ACK needs to be moved earlier for sure.

Reply:
> > Hrm, true. I think the definition of the ACK is not that clear.
> 
> Downlink and uplink should not be handled in a different way on that level. IPC4 is a one thing, but definition of the sideband IPC behavior is something else and semantics of the operations on symmetric unidirectional communication interfaces should be consistent. When the DSP is the initiator, it may send a new message anytime the communication channel is free. The same should be assumed for any other IPC channel no matter if the DSP is the initiator or the target.

The SOF (IPC3) IPC implementation does not have ACK, there is only one step: reply.
IPC4 on the other hand uses 'split' protocol: ACK (which is the reply in SOF.IPC3), and empty notification to host that the message is under processing and DSP can accept a new message. The reply is a new firmware initiated message (like SOF.IPC3 notification or IPC4 notification).

In IPC3 case we don't send reply message as the reply is the ACK of the reception of the message, sent when all processing is done.

I think this difference merits a bit different handling of the same IPC registers, their use is the same, but the protocol is different on top of them.

> 
> > The kernel implementation [...]
> 
> FW side implementation should not depend on this particular implementation of the host side handler as the IPC4 protocol is supposed to be used by other drivers as well.

Sure, it should not, it was just a note that the Linux kernel is serializing messages, never have more than one message sent at the same time, we wait for the reply.

Reply:
Note, that currently we do rely on host-side IPC serialisation, i.e. on the fact, that the host doesn't send new IPCs until we've completed and replied to the previous one. There are even several comments in the code about that, next to location that we know depend on that serialisation, and there are definitely more of them unmarked as such.